iterative-walkthrough/deployment-2/guestbook-redis/main.go

```diff
@@ -1,74 +1,111 @@
 package main
 
 import (
-    "log"
-    "net/http"
-    "os"
-    "github.com/zpatrick/fireball"
-    "github.com/garyburd/redigo/redis"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+
+	"github.com/garyburd/redigo/redis"
+	"github.com/hashicorp/consul/api"
+	"github.com/zpatrick/fireball"
 )
 
 var (
-    redisProtocol = "tcp"
-    redisAddress = os.Getenv("REDIS_ADDRESS_AND_PORT")
+	redisProtocol       = "tcp"
+	redisAddressAndPort = os.Getenv("REDIS_ADDRESS_AND_PORT")
 )
 
+func getRedisAddress() (string, error) {
+	if redisAddressAndPort != "" {
+		return redisAddressAndPort, nil
+	}
+
+	c := api.Catalog{}
+	services, _, err := c.Service("redis", "", nil)
+	if err != nil {
+		return "", err
+	}
+
+	for _, service := range services {
+		if service.ServiceName == "redis" {
+			return fmt.Sprintf("%s:%d", service.ServiceAddress, service.ServicePort), nil
+		}
+	}
+
+	return "", fmt.Errorf("service 'redis' could not be found.")
+}
+
 func listEntries(c *fireball.Context) (fireball.Response, error) {
-    conn, err := redis.Dial(redisProtocol, redisAddress)
-    if err != nil {
-        log.Fatal(err)
-    }
-    defer conn.Close()
-
-    entries, _ := redis.Strings(conn.Do("LRANGE", "entries", 0, -1))
-    return c.HTML(200, "index.html", entries)
+	redisAddress, err := getRedisAddress()
```

Comment:
It seems like calling `getRedisAddress` is unnecessary in these various places if you instead ran `getRedisAddress` in main and mutated the `redis` connection params